Parse error: Field access of compound literal

The following works in GCC and clang; it seems to be the only error preventing the frontend from parsing `zig2.c`:

```c
struct B {
   _Bool payload;
};

int foo(void) {
   return (struct B) {.payload = 1}.payload;
}
```

```
./test.c:6:36: error: expected ';', found '.'
   return (struct B) {.payload = 1}.payload;
                                   ^
```

--- a/src/Parser.zig
+++ b/src/Parser.zig
@@ -5667,18 +5667,9 @@ fn castExpr(p: *Parser) Error!Result {
         try p.expectClosing(l_paren, .r_paren);
 
         if (p.tok_ids[p.tok_i] == .l_brace) {
-            // compound literal
-            if (ty.isFunc()) {
-                try p.err(.func_init);
-            } else if (ty.is(.variable_len_array)) {
-                try p.err(.vla_init);
-            } else if (ty.hasIncompleteSize() and !ty.is(.incomplete_array)) {
-                try p.errStr(.variable_incomplete_ty, p.tok_i, try p.typeStr(ty));
-                return error.ParsingFailed;
-            }
-            var init_list_expr = try p.initializer(ty);
-            try init_list_expr.un(p, .compound_literal_expr);
-            return init_list_expr;
+            // Compound literal; handled in unExpr
+            p.tok_i = l_paren;
+            break :cast_expr;
         }
 
         var operand = try p.castExpr();
@@ -5867,7 +5858,7 @@ fn offsetofMemberDesignator(p: *Parser, base_ty: Type) Error!Result {
 }
 
 /// unExpr
-///  : primaryExpr suffixExpr*
+///  : (compoundLiteral | primaryExpr) suffixExpr*
 ///  | '&&' IDENTIFIER
 ///  | ('&' | '*' | '+' | '-' | '~' | '!' | '++' | '--' | keyword_extension | keyword_imag | keyword_real) castExpr
 ///  | keyword_sizeof unExpr
@@ -6192,8 +6183,11 @@ fn unExpr(p: *Parser) Error!Result {
             return operand;
         },
         else => {
-            var lhs = try p.primaryExpr();
-            if (lhs.empty(p)) return lhs;
+            var lhs = try p.compoundLiteral();
+            if (lhs.empty(p)) {
+                lhs = try p.primaryExpr();
+                if (lhs.empty(p)) return lhs;
+            }
             while (true) {
                 const suffix = try p.suffixExpr(lhs);
                 if (suffix.empty(p)) break;
@@ -6202,6 +6196,30 @@ fn unExpr(p: *Parser) Error!Result {
             return lhs;
         },
     }
+}
+
+/// compoundLiteral
+///  : '(' type_name ')' '{' initializer_list '}'
+///  | '(' type_name ')' '{' initializer_list ',' '}'
+fn compoundLiteral(p: *Parser) Error!Result {
+    const l_paren = p.eatToken(.l_paren) orelse return Result{};
+    const ty = (try p.typeName()) orelse {
+        p.tok_i = l_paren;
+        return Result{};
+    };
+    try p.expectClosing(l_paren, .r_paren);
+
+    if (ty.isFunc()) {
+        try p.err(.func_init);
+    } else if (ty.is(.variable_len_array)) {
+        try p.err(.vla_init);
+    } else if (ty.hasIncompleteSize() and !ty.is(.incomplete_array)) {
+        try p.errStr(.variable_incomplete_ty, p.tok_i, try p.typeStr(ty));
+        return error.ParsingFailed;
+    }
+    var init_list_expr = try p.initializer(ty);
+    try init_list_expr.un(p, .compound_literal_expr);
+    return init_list_expr;
 }
 
 /// suffixExpr

--- a/test/cases/compound literals.c
+++ b/test/cases/compound literals.c
@@ -1,0 +1,29 @@
+struct A {
+    int x;
+};
+struct B {
+    struct A a;
+    int y;
+};
+
+int *p = (int []){2, 4};
+
+void foo(struct A *a) {
+    (void)a;
+}
+void bar(void) {
+    int a = (struct B){
+        .y = 1,
+        .a = {.x = 2},
+    }.a.x;
+    int b = (int []){1,2,3,4,5}[4];
+    int c = (int []){1,2,3,4,5}[10];
+    foo(&(struct A){.x = 42});
+
+    char *str1 = "string 1";
+    char *str2 = (char []){"string 2"};
+    const char *str3 = (const char []){"string 3"};
+}
+
+#define EXPECTED_ERRORS "compound literals.c:20:32: warning: array index 10 is past the end of the array [-Warray-bounds]" \
+

